packages/google_sign_in/lib/src/google_sign_in.dart

```diff
@@ -0,0 +1,302 @@
+    if (_user != null) {
+      return _user;
+    }
+
+    final AuthorizationResponse authorizationResponse =
+        await _authClient.requestAuthorization(
+      configuration.clientId,
+      configuration.scope,
+    );
+
+    final Future<TokenResponse?> tokenResponseFuture = _authClient.pollToken(
+      clientId: configuration.clientId,
+      clientSecret: configuration.clientSecret,
+      deviceCode: authorizationResponse.deviceCode,
+      interval: authorizationResponse.interval,
+    );
+
+    showDeviceFlowWidget(
+      code: authorizationResponse.userCode,
+      verificationUrl: authorizationResponse.verificationUrl,
+      expiresIn: authorizationResponse.expiresIn,
+      onExpired: () => _authClient.cancelPollToken(),
+      onCancelled: () => _authClient.cancelPollToken(),
+    );
+
+    // Waits until user interaction on secondary device is finished, or until
+    // code is expired or polling is cancelled.
+    final TokenResponse? tokenResponse = await tokenResponseFuture;
+    if (tokenResponse == null) {
+      return null;
+    }
+    closeDeviceFlowWidget();
+
+    final Map<String, dynamic> jsonProfile =
+        utils.decodeJWT(tokenResponse.idToken);
+    final ProfileData profile = ProfileData.fromJson(jsonProfile);
+
+    final Authentication authentication = Authentication(
+      clientId: configuration.clientId,
+      accessToken: tokenResponse.accessToken,
+      accessTokenExpirationDate: DateTime.now().add(tokenResponse.expiresIn),
+      refreshToken: tokenResponse.refreshToken,
+      idToken: tokenResponse.idToken,
+    );
+
+    utils.checkFormat<String>(<String>['sub'], jsonProfile);
+    _user = GoogleUser(
+      userId: jsonProfile['sub'] as String,
+      profile: profile,
+      authentication: authentication,
+      scope: tokenResponse.scope,
+    );
+
+    return _user;
+  }
+
+  /// Returns the [Authentication] of the currently signed in user, returns
+  /// `null` if no user is signed in.
+  ///
+  /// If [refresh] is `true`, expired tokens will be refreshed after
+  /// getting new tokens from the server. [clientSecret] cannot be `null` when
+  /// refresh is required.
+  Future<Authentication?> getAuthentication({
+    bool refresh = false,
+    String? clientSecret,
+  }) async {
+    if (_user == null) {
+      return null;
+    }
+    final GoogleUser user = _user!;
+    final Authentication authentication = user.authentication;
+
+    // Check access token is if expired.
+    const Duration minimalTimeToExpire = Duration(minutes: 1);
+    if (!refresh ||
+        authentication.accessTokenExpirationDate
+            .add(minimalTimeToExpire)
+            .isBefore(DateTime.now())) {
+      return authentication;
+    }
+
+    if (clientSecret == null) {
+      throw ArgumentError(
+          "Token expired: `clientSecret` can't be null when refresh is required",
+          'clientSecret');
+    }
+
+    if (authentication.refreshToken == null) {
+      throw PlatformException(
```

Comment:
`PlatformException` should literally indicate a platform error (platform API failure in most cases).

Reply:
My decision of using `PlatformException` came from [google_sign_in_ios](https://github.com/flutter/plugins/blob/160967b311d6b47a9aee032071c1cb560a43e196/packages/google_sign_in/google_sign_in_ios/lib/google_sign_in_ios.dart#L33-L37), it views that dart implementation of platform channel is also considered "platform error". However, I can switch it to another custom exception class (something like `GoogleSignInTizenException`) or built-in exceptions (though I'm not sure which ones to use yet) if that makes more sense in our case.

Reply:
Okay. `PlatformException` looks better than a custom exception type or plain `Exception`.